src/pydsl/triton.py

```diff
@@ -0,0 +1,239 @@
+                callee_name = (
+                    op.sym_name.value + prefix + str(len(signatures))
+                )  # add a number to the end of the function name
+                if current_signature in signatures.keys():
+                    callee_name = (
+                        op.sym_name.value
+                        + prefix
+                        + str(list(signatures.keys()).index(current_signature))
+                    )
+                    callee = signatures[current_signature]
+                    continue  # this function has already been compiled
+                op.attributes["sym_name"] = mlirir.StringAttr.get(callee_name)
+                new_op = op.clone()
+                callee = new_op
+                remove_uneccessary_args(callee)
+                signatures[current_signature] = callee
+            else:
+                op.clone()
+    return callee
+
+
+def run_and_get_output_stdin(
+    cmds: Iterable[str], stdin_string: str
+) -> str | None:
+    result = subprocess.run(
+        " ".join(cmds),
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        input=stdin_string.encode("utf-8"),
+        shell=True,
+        check=True,
+    )
+    return result.stdout.decode("utf-8") if result.stdout else None
+
+
+def triton_to_linalg(
+    module: mlirir.Module,
+    triton_func: JITFunction,
+    args: list[tuple[str, typing.Any]],
+    triton_funcs: list,
+) -> func.FuncOp:
+    """
+    Converts triton functions to linalg dialect using triton-adapter-opt.
+    args is a list of tuples of the form (name, type), where name is the
+    name of the triton function argument and type is the PyDSL type.
+    returns the converted function as a func.FuncOp.
+    """
+
+    signature = {}
+    constexprs = {}
+    for name, typ in args:
+        signature[name] = pydsl_type_to_triton_str(type(typ))
+        if isinstance(typ, Number):  # handling the constexpr case
+            constexprs[name] = typ.value
+    triton_src = triton.compiler.ASTSource(
+        fn=triton_func,
+        signature=signature,
+        constexprs=constexprs,
+    )
+    # Fixes the backend for CUDA, so that it can be used with triton-adapter-opt.
+    # 80 is the gpu architecture (i.e. sm_80), 32 is the warp size.
+    # These particular values aren't important, but they need to be set to some
+    # valid CUDA backedn for triton-adapter-opt to work.
+    target = GPUTarget("cuda", 80, 32)
+    triton_mlir = triton.compile(triton_src, target=target).asm["ttir"]
+
+    mlir_module_str = run_and_get_output_stdin(
+        ["triton-adapter-opt", "--triton-to-linalg"], triton_mlir
+    )
+    triton_module = mlirir.Module.parse(mlir_module_str)
+    current_signature = " ".join([
+        str(v) + "=" + str(constexprs[k]) if k in constexprs else str(v)
+        for k, v in signature.items()
+    ])
+    func_op = combine_mlir_modules(
+        module, triton_module, triton_funcs, current_signature
+    )
```

Comment:
```suggestion
    signature_str = " ".join([
        f"{typ_str}={str(constexprs[name])}" if name in constexprs else typ_str
        for name, typ_str in signature.items()
    ])
    func_op = combine_mlir_modules(
        module
```
`typ_str` is a much better name and it does not need to be cast to `str` again. The f-string I added so it fits in line width with my changes, I don't know whether it's better in general than string concatenation.

In general, I think converting constexpr to str is a bit suspicious. What if the constexpr is a string containing a space? E.g. `kernel("xyz i32")` and `kernel("xyz", UInt32(123))` would have the same signature under this system. Maybe it's best not to serialize at all?